pkg/controllers/updaterun/stop.go

```diff
@@ -0,0 +1,252 @@
+		if condition.IsConditionStatusTrue(meta.FindStatusCondition(curCluster.Conditions, string(placementv1beta1.ClusterUpdatingConditionSucceeded)), updateRun.GetGeneration()) {
+			// The cluster status is marked as deleted.
+			unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the deleted cluster `%s` in the deleting stage still has a binding", bindingSpec.TargetCluster))
+			klog.ErrorS(unexpectedErr, "The cluster in the deleting stage is not removed yet but marked as deleted", "cluster", curCluster.ClusterName, "updateRun", updateRunRef)
+			return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+		}
+		if condition.IsConditionStatusTrue(meta.FindStatusCondition(curCluster.Conditions, string(placementv1beta1.ClusterUpdatingConditionStarted)), updateRun.GetGeneration()) {
+			// The cluster status is marked as being deleted.
+			if binding.GetDeletionTimestamp().IsZero() {
+				// The cluster is marked as deleting but the binding is not deleting.
+				unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the cluster `%s` in the deleting stage is marked as deleting but its corresponding binding is not deleting", curCluster.ClusterName))
+				klog.ErrorS(unexpectedErr, "The binding should be deleting before we mark a cluster deleting", "clusterStatus", curCluster, "updateRun", updateRunRef)
+				return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+			}
+			continue
+		}
+	}
+
+	// The rest of the clusters in the stage are not in the toBeDeletedBindings so it should be marked as delete succeeded.
+	for _, clusterStatus := range existingDeleteStageClusterMap {
+		// Make sure the cluster is marked as deleted.
+		if !condition.IsConditionStatusTrue(meta.FindStatusCondition(clusterStatus.Conditions, string(placementv1beta1.ClusterUpdatingConditionStarted)), updateRun.GetGeneration()) {
+			markClusterUpdatingStarted(clusterStatus, updateRun.GetGeneration())
+		}
+		markClusterUpdatingSucceeded(clusterStatus, updateRun.GetGeneration())
+	}
+
+	klog.V(2).InfoS("The delete stage is stopping", "numberOfDeletingClusters", len(toBeDeletedBindings), "updateRun", updateRunRef)
+	allDeletingClustersDeleted := true
+	for _, clusterStatus := range updateRunStatus.DeletionStageStatus.Clusters {
+		klog.InfoS("Checking deletion status of cluster in delete stage", "cluster", clusterStatus.ClusterName, "conditions", clusterStatus.Conditions, "updateRun", updateRunRef)
+		if condition.IsConditionStatusTrue(meta.FindStatusCondition(clusterStatus.Conditions,
+			string(placementv1beta1.ClusterUpdatingConditionStarted)), updateRun.GetGeneration()) && !condition.IsConditionStatusTrue(
+			meta.FindStatusCondition(clusterStatus.Conditions, string(placementv1beta1.ClusterUpdatingConditionSucceeded)),
+			updateRun.GetGeneration()) {
+			allDeletingClustersDeleted = false
+			break
+		}
+	}
+
+	if allDeletingClustersDeleted || len(toBeDeletedBindings) == 0 {
```

Comment:
Do we need to check if len(toBeDeletedBindings), if len(toBeDeletedBindings) == 0 wouldn't allDeletingClustersDeleted be true ?